src/Microsoft.Azure.NotificationHubs/Messaging/AuthorizationRules.cs

```diff
@@ -0,0 +1,301 @@
+    {
+        /// <summary>
+        /// The serializer
+        /// </summary>
+        public static readonly DataContractSerializer Serializer = new DataContractSerializer(typeof(AuthorizationRules));
+        /// <summary>
+        /// The inner collection
+        /// </summary>
+        public readonly ICollection<AuthorizationRule> innerCollection;
+        readonly IDictionary<string, SharedAccessAuthorizationRule> nameToSharedAccessAuthorizationRuleMap;
+        bool duplicateAddForSharedAccessAuthorizationRule;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="T:Microsoft.Azure.NotificationHubs.Messaging.AuthorizationRules"/> class.
+        /// </summary>
+        public AuthorizationRules()
+        {
+            this.nameToSharedAccessAuthorizationRuleMap = new Dictionary<string, SharedAccessAuthorizationRule>(StringComparer.OrdinalIgnoreCase);
+            this.innerCollection = new List<AuthorizationRule>();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="T:Microsoft.Azure.NotificationHubs.Messaging.AuthorizationRules"/> class with a list of <see cref="T:Microsoft.Azure.NotificationHubs.Messaging.AuthorizationRule"/>.
+        /// </summary>
+        /// <param name="enumerable">The list of <see cref="T:Microsoft.Azure.NotificationHubs.Messaging.AuthorizationRule"/>.</param>
+        public AuthorizationRules(IEnumerable<AuthorizationRule> enumerable)
+        {
+            if (enumerable == null)
+            {
+                throw new ArgumentNullException("enumerable");
+            }
+
+            this.nameToSharedAccessAuthorizationRuleMap = new Dictionary<string, SharedAccessAuthorizationRule>(StringComparer.OrdinalIgnoreCase);
+            this.innerCollection = new List<AuthorizationRule>();
+
+            foreach (AuthorizationRule rule in enumerable)
+            {
+                this.Add(rule);
+            }
+        }
+
+        /// <summary>
+        /// Returns an enumerator that iterates through the collection.
+        /// </summary>
+        /// <returns>
+        /// An enumerator that can be used to iterate through the collection.
+        /// </returns>
+        public IEnumerator<AuthorizationRule> GetEnumerator()
+        {
+            return this.innerCollection.GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return ((IEnumerable)this.innerCollection).GetEnumerator();
+        }
+
+        /// <summary>
+        /// Adds the specified <see cref="T:Microsoft.Azure.NotificationHubs.Messaging.AuthorizationRule"/> into the collection.
+        /// </summary>
+        /// <param name="item">The <see cref="T:Microsoft.Azure.NotificationHubs.Messaging.AuthorizationRule"/> to be added.</param>
+        public void Add(AuthorizationRule item)
+        {
+            if (item is SharedAccessAuthorizationRule)
```

Comment:
You could collapse this into `if (item is SharedAccessAuthorizationRule newRule)`